test/language/languageUnderstandingTest.js

```diff
@@ -1,40 +1,276 @@
+            })
+            .catch((err) => {
+                if (err.code === 'ETIMEDOUT') {
+                    retry(err);
+                }
+                throw err;
+            });
+        });  
+    }  
+    var createTrainPublishApp = () => {
+
+        client.isTrained = isTrained;
+        let body = {
+            "domainName": "Web", 
+            "culture": "en-us"
+        };
+        return client.addPrebuiltDomain(body).then(results =>{
+            client.appID = results.substring(results.length-36, results.length);
+            console.log(client.appID);
+            return client.train();
+        }).then(results => {
+            return waitUntilTrained(client);
+        }).then((response) => {
+            return client.publish({
+                    "versionId": "0.1",
+                    "isStaging": false,
+                    "region": "westus"
+                    });
+        }).then((response) => {
+            return response;
+        }).catch(err => {
+            console.log(err);
+            throw(err);
+        });        
+    }
+    var deleteTestApp = () =>{
+        return client.deleteApp()
+        .then((response) => {
+            response.should.not.be.undefined();
+            client.appID = undefined;
+            return response;
+        }).catch((err) => {
+            throw(err);
+        });
+    }
+    describe("create app before, don't delete after", () => {
+        before((done) => {
+            createTrainPublishApp()
+            .then(results => {
+                done();
+            }).catch(err => {
+                done(err);
+            });
+        });
+        it('should delete app', (done) => {   
+            client.deleteApp()
+            .then((response) => {
+                response.should.not.be.undefined();
+                response.should.have.properties(['code', 'message']);
+                response.code.should.equal("Success");
+                response.message.should.equal("Operation Successful");
+                done();
+            }).catch((err) => {
+                done(err);
+            });
+        })         
+    })
+    describe("Does not create app, no need to delete", () => {
+        it('should get list of applications', (done) => {
+
+            let parameters = {
+                skip:0,
+                take:100
+            };
+
+            client.getApplications(parameters)
+            .then((response) => {
+                response.should.not.be.undefined();
+                response.should.be.Array;
+                if (response.length > 0) {
+                    response[0].should.have.properties(['id', 'name', 'description','culture','usageScenario','domain','versionsCount','createdDateTime','endpoints','endpointHitsCount','activeVersion']);
+                }
+                done();
+            }).catch((err) => {
+                done(err);
+            });
+        })
+    })
+    describe("Delete app after", () => {
```

Comment:
Are all 4 of these `describe` blocks really necessary? Can we merge 1,2,4 into one?

Reply:
The inner describe blocks denote when an before, after, or both is used in creating the app and deleting the app used in the test. Is that what you are referring to with this comment?

Reply:
Sorry, I was not clear. The 4 inner `describe` blocks give me the feeling that we're creating and deleting too many apps which can be expensive & slow.

I'd like to see only two inner `describe`s:

- One that creates an app beforehand:
   - `before`: creates, trains and publishes 1 app
   - runs the tests (lists app versions, exports app versions, returns queries, get labeled examples, detect intent, user access list, version entities, etc...)
   - `after`: deletes the app
- One that does not create an app beforehand because the tests are creating them:
   - `afterEach` deletes the app
   - runs the tests: import app

Your helper functions `waitUntilTrained`, `createTrainPublishApp` and `deleteTestApp` are already testing many things so there's no need to have specific tests for these: get train status, add prebuilt domain, train, publish, delete app. We can move the specific response checks to each of these functions.